### ISSUE TITLE
Separate Material properties and internal state variables

### ISSUE THREAD

**Issue by jvaara** [JuliaFEM/Materials.jl#34]
Currently the Material base struct looks like this:
```
mutable struct Material{M<:AbstractMaterial}
    stress :: Vector{Float64}
    strain :: Vector{Float64}
    dstress :: Vector{Float64}
    dstrain :: Vector{Float64}
    jacobian :: Matrix{Float64}
    time :: Float64
    dtime :: Float64
    properties :: M
end
```
where `properties` holds both material specific parameters (read from fields and material does not change them) as well as the internal state variables (stored in integration points and material integration needs them to keep track of its state). 

Separating these would make sense for in the future if someone wants to automatize e.g. `material_preprocess_analysis`, `material_preprocess_increment` and `material_postprocess_increment`. Also keeping track of changes in parameters is something we need to look at in some point.

**Comment by jvaara** [comment]
# Requirements for new `AbstractMaterial` structure:
## Categorization of the data by functionalities
`variables` = **previous converged values of the internal state variables (variables that defines the material state and material has control over)** (e.g. `stress`, `plastic_strain`,...). `dvariables` to be defined by the `integrate_material!` function over the increment. `dvariables` are iterated until global convergence is obtained. Converged `variables` values should be stored in integration points. Structure has to support automatic generation of `material_postprocess_increment!` for any material (e.g. `variables` += `dvariables`).

`drivers` = **variables that drive evolution of the material state (variables that material has no direct control over)** (e.g. `strain`, `time`, `temperature`...). `ddrivers` to define the increment of `drivers` at the beginning of increment/iteration by global solver. If `ddrivers = 0`, then `dvariables = 0`.

`parameters` = **parameter values at the beginning of the increment** (e.g. `youngs_modulus`, `nu`, `yield_stress`). These parameters are used by the evolution equations of the material model. `dparameters` to capture changes in `parameters` during the increment. These should be `field`s in JuliaFEM and interpolated to the integration points at `material_preprocess_increment!`. Naming of these would preferably be same in the material model as in the fields, so automatical interpolation mapping can be performed (either `String` with no white spaces or `Symbol`).

(`source_terms` or `fluxes` (e.g. plastic dissipation generating heat, piezoelectric material deformation producing current - defined by the material model) - possibly to be neglected at this moment)

## JuliaFEM bindings

The interface to JuliaFEM is at the moment defined with 7 functions
- `integrate_material!(material)`: Defines the material behaviour over time increment `dvariables` as well as interaction with the global residual with `source_terms`/`fluxes`, given `variables`, `parameters`, `dparameters` and `ddrivers`.
- `material_preprocess_analysis!(material, element, ip, time)`: Initializes `ip` for data storage of both `variables` and `drivers` at simulation start `time`.
- `material_preprocess_increment!(material, element, ip, time)`: Looks for `parameters` from `element` `field`s to `ip`s at `time` (end time of the increment) and notices if the `material` `parameters` at previous converged state differ, then sets `dparameters`. Updates `ddrivers` that are not iterated by the global solver (e.g. `dtime`).
- `material_preprocess_iteration!(material, element, ip, time)`: Updates `ddrivers` that are iterated by the global solver over the increment (e.g. `dstrain`).
- `material_postprocess_iteration!(material, element, ip, time)`: This could be used to pick the trial values of `dvariables` and `source_terms`/`fluxes` for the global solver residual check. At the moment this is defined by the `Problem` (i.e. what is the global solved variable/physics) and thus not used that often.
- `material_postprocess_increment!(material, element, ip, time)`: Updates the converged state (`variables += dvariables`, `parameters += dparameters`, `drivers += ddrivers`) and resets the increments to (`dvariables = 0`, `dparameters = 0`, [`ddrivers = 0`]).
- `material_postprocess_analysis!(material, element, ip, time)`: No use for this has been found yet :)

## Other remarks

- Anything can be of any size (and preferably type for possibility to implement with `Tensor` as well as `Vector`/`Matrix`).
- Structure to enable automatization of the JuliaFEM and other simulator bindings.

**Comment by KristofferC** [comment]
For the state, I think something like the following would be nice:

```
abstract type AbstractMaterialState end

@generated function Base.:+(state::T, dstate::T) where {T <: AbstractMaterialState}
   expr = [:(state.$p+ dstate.$p) for p in fieldnames(T)]
   return :(T($(expr...)))
end
```

It looks a bit complicated but it basically just means that any struct that subtypes `AbstractMaterialState` can be added and the result is such that all fields are added. For example:

```
julia> struct SomeState <: AbstractMaterialState
           sigma::Float64
           eps_p::Float64
       end

julia> state = SomeState(1.0, 2.0)
SomeState(1.0, 2.0)

julia> dstate = SomeState(0.1, 0.2)
SomeState(0.1, 0.2)

julia> state + dstate
SomeState(1.1, 2.2)
```

We can also define more methods on `AbstractMaterialState`, for example:

```
abstract type AbstractMaterialState end

for f in (:+, :-)
   @eval begin
       @generated function Base.$(f)(state::T, dstate::T) where {T <: AbstractMaterialState}
          expr = [:($($f)(state.$p, dstate.$p)) for p in fieldnames(T)]
          return :(T($(expr...)))
       end
   end
end

for f in (:*, :/)
   @eval begin
       @generated function Base.$(f)(state::T, num::Number) where {T <: AbstractMaterialState}
          expr = [:($($f)(state.$p, num)) for p in fieldnames(T)]
          return :(T($(expr...)))
       end
   end
end

Base.:*(num::Number, state::T) where {T <: AbstractMaterialState} = state * num

@generated function Base.zero(state::T) where {T <: AbstractMaterialState}
    expr = [:(zero(state.$p)) for p in fieldnames(T)]
    return :(T($(expr...)))
end
```

and then we can do:

```
julia> zero(SomeState)
SomeState(0.0, 0.0)

julia> state = SomeState(1.0, 2.0)
SomeState(1.0, 2.0)

julia> state * 2
SomeState(2.0, 4.0)

julia> state / 3
SomeState(0.3333333333333333, 0.6666666666666666)

julia> state - state
SomeState(0.0, 0.0)
```

So if we want to integrate a state over the whole domain, it would be quite simple

**Comment by KristofferC** [comment]
As I posted in Gitter, doing `+` and returning a new material state costs a little bit if we are wrapping non-`isbits´ object in the state, as an example:

```
using Tensors 

struct MatStatTensor <: AbstractMaterialState
    t::Vec{3, Float64}
end

struct MatStatVector <: AbstractMaterialState
    t::Vector{Float64}
end

stat_tensors = [MatStatTensor(rand(Vec{3})) for i in 1:10^5];
stat_vector =  [MatStatVector(rand(3)) for i in 1:10^5];
```

Here the first struct using the `Tensor` object is stack allocated while the normal vector is (of course) heap allocated. The result is

```
julia> @time sum(stat_tensors)
  0.000374 seconds (5 allocations: 192 bytes)
MatStatTensor([49994.0, 49925.9, 49946.1])

julia> @time sum(stat_vector)
  0.012497 seconds (200.00 k allocations: 12.207 MiB, 43.76% gc time)
MatStatVector([49980.9, 50046.0, 50055.1])
```

So perhaps it is worth adding `mutable` and allowing things like `add!(state, dstate)` updating the `state`?

**Comment by KristofferC** [comment]
My main point here is that I definitely think that keeping all the states in one struct and having separate structs for `state` and `dstate` is better than the `Variable{T}` with `state` and `dstate` variables in it.

**Comment by jvaara** [comment]
I performed a couple of benchmarks of the two approaches in [Dual variables here](https://github.com/JuliaFEM/Materials.jl/blob/abstractmaterial/examples/abstractmaterial_testing.jl) and [States here](https://github.com/JuliaFEM/Materials.jl/blob/abstractmaterial/examples/abstractmaterial_testing_states.jl)

Benchmarking with `Chaboche`-style material model gave following results:
*"Dual Variable" approach:
```
Benchmark Chaboche VariableState
  609.608 µs (2987 allocations: 109.73 KiB)
```
*State approach:
```
Benchmark Chaboche State
  183.525 µs (1 allocation: 336 bytes)
```
Here the `state`-approach was superior. If the `randn(SymmetricTensor{2,3})` structure wasn't taken advantage of (more realistic case?), then the results were as follows:

*"Dual variable" approach:
```
Benchmark Chaboche VariableState
  775.014 µs (7987 allocations: 734.73 KiB)
```
*State approach:
```
Benchmark Chaboche State
  540.347 µs (25985 allocations: 1.72 MiB)
```

The use of these different styles is probably highlighted in how the code looks:
```
state.stress.dvalue = randn(SymmetricTensor{2,3})
state.strain.dvalue = randn(SymmetricTensor{2,3})
state.backstress1.dvalue = randn(SymmetricTensor{2,3})
state.backstress2.dvalue = randn(SymmetricTensor{2,3})
state.plastic_strain.dvalue = randn(SymmetricTensor{2,3})
state.cumeq.dvalue = norm(state.plastic_strain.dvalue)
state.R.dvalue = randn()
update!(state) # To be called outside of material integration
```
versus 
```
dstress = randn(SymmetricTensor{2,3})
dstrain = randn(SymmetricTensor{2,3})
dbackstress1 = randn(SymmetricTensor{2,3})
dbackstress2 = randn(SymmetricTensor{2,3})
dplastic_strain = randn(SymmetricTensor{2,3})
dcumeq = norm(dplastic_strain)
dR = randn()
dstate = AnotherState(dstress, dstrain, dbackstress1,
                           dbackstress2, dplastic_strain, dcumeq, dR)
state = state + dstate # To be called outside of material integration
```
In my eyes, the second (state) approach is probably easier to adopt for developers with Matlab/UMAT(Fortran) background.

**Comment by jvaara** [comment]
Old style material benchmark using Arrays and view [here](https://github.com/JuliaFEM/Materials.jl/commit/3f3a44fcac24c2080f85afc3a5437904988506ea)
```
    Benchmark Chaboche Old State
      776.307 µs (20011 allocations: 2.06 MiB)
```

So as discussed in Gitter, it probably boils down to the ease of automatizing the `material_postprocess_increment!` and other similar methods from the hierarchy.

**Comment by jvaara** [comment]
Closed with Merge #45
